builder/image_solver.go

```diff
@@ -227,33 +205,35 @@ func (s *localRegistryImageSolver) SolveImage(ctx context.Context, mts *states.M
 		return nil
 	}
 
-	var saveImage = mts.Final.LastSaveImage()
-	imgJSON, err := json.Marshal(saveImage.Image)
-	if err != nil {
-		return nil, errors.Wrap(err, "image json marshal")
-	}
+	buildFn := func(childCtx context.Context, gwClient gwclient.Client) (*gwclient.Result, error) {
+		res := gwclient.NewResult()
 
-	if !strings.Contains(dockerTag, ":") {
-		dockerTag += ":latest"
-	}
+		for i, imageDef := range imageDefs {
+			var saveImage = imageDef.MTS.Final.LastSaveImage()
+			imgJSON, err := json.Marshal(saveImage.Image)
+			if err != nil {
+				return nil, errors.Wrap(err, "image json marshal")
+			}
 
-	bf := func(childCtx context.Context, gwClient gwclient.Client) (*gwclient.Result, error) {
-		ref, err := llbutil.StateToRef(childCtx, gwClient, saveImage.State, true, mts.Final.PlatformResolver, s.cacheImports.AsMap())
-		if err != nil {
-			return nil, errors.Wrap(err, "initial state to ref conversion")
-		}
+			if !strings.Contains(imageDef.ImageName, ":") {
+				imageDef.ImageName += ":latest"
+			}
 
-		res := gwclient.NewResult()
+			ref, err := llbutil.StateToRef(childCtx, gwClient, saveImage.State, true, imageDef.MTS.Final.PlatformResolver, m.cacheImports.AsMap())
```

Comment:
Would calling this in parallel (for each `imageDef`) speed things up?

Reply:
actually... probably not, as we're not forcing a solve via `ref.ReadFile/Dir/Stat` (since we just return the ref via https://github.com/earthly/earthly/blob/71ea6dcca8b4590b7ee4338f78a8c2646ccc3e49/util/llbutil/statetoref.go#L44 )

ignore this :)

Reply:
Hmm. I suppose it still does a `Solve` request for each image in `StateToRef`. I can try making this concurrent to see if there's a noticeable performance change.

Reply:
It doesn't seem to account for much (~300ms) but it doesn't seem to be hurting either.

Reply:
the initial call to `Solve` just gets it queued up in buildkit's solver/scheduler, and the returned ref will point to a state that might not be ready (solved) yet. so it makes sense that it's only 300ms (since it's just a tcp-round trip to kick off the build), it's not until we try to read from the ref which is when things actually block. but if it speeds it up and doesn't break, sounds good :)